src/engine.ml

```diff
@@ -951,31 +951,52 @@ let incoming_control_server auth_user_pass is_not_taken config rng session
       let ciphers = tls_ciphers config and version = tls_version config in
       let* authenticator =
         match
-          (Config.find Verify_client_cert config, Config.find Ca config)
+          ( Config.find Verify_client_cert config,
+            Config.find Ca config,
+            Config.find Peer_fingerprint config )
         with
-        | None, Some ca | Some `Required, Some ca ->
+        | None, Some ca, None | Some `Required, Some ca, None ->
             Ok
               (Some
                  (X509.Authenticator.chain_of_trust
                     ~time:(fun () -> Some now)
                     ~allowed_hashes:Mirage_crypto.Hash.hashes ca))
-        | Some `None, _ ->
+        | None, None, Some fps | Some `Required, None, Some fps ->
+            Ok
+              (Some
+                 (fun ?ip ~host chain ->
+                   List.fold_left
+                     (fun acc fingerprint ->
+                       match acc with
+                       | Ok _ -> acc
+                       | Error _ ->
+                           X509.Validation.trust_cert_fingerprint
+                             ~time:(fun () -> Some now)
+                             ~hash:`SHA256 ~fingerprint ?ip ~host chain)
+                     (Error (`Msg "No fingerprints provided"))
+                     fps))
+        | Some `None, _, _ ->
             Log.warn (fun m ->
                 m
                   "server with 'verify-client-cert none', ensure a different \
                    authentication mechanism is set up.");
```

Comment:
We should log this warning in verify_server_config (or how the function is named). Otherwise it's logged on very new connection rather than just once.

Reply:
done in 1a851ea